### PR TITLE
HDDS-10192. Verify container checksum after downloaded

### PR DESCRIPTION
**Pull request by DaveTeng0**
## What changes were proposed in this pull request?
HDDS-10192 Verify container checksum after downloaded

Verify container checksum after downloaded to make sure the checksum matches the value stored in yaml;otherwise, throw error immediately.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-10192

## How was this patch tested?
unit test.

### CODE REVIEW

**Comment by DaveTeng0** [comment]
cc. @sumitagrawl @errose28 @jojochuang @umamaheswararao @nandakumar131 @sadanand48

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/helpers/TestContainerUtils.java [hunk 1]
```diff
@@ -122,6 +127,23 @@ public void testDatanodeIDPersistent(@TempDir File tempDir) throws Exception {
...
+    // mock to return different checksum
+    when(containerData.getChecksum()).thenReturn("checksum1", "checksum2");
+    conf.setBoolean(HddsConfigKeys.HDDS_CONTAINER_CHECKSUM_VERIFICATION_ENABLED, true);
+    try {
+      ContainerUtils.verifyChecksum(containerData, conf);
+      fail("code should not reach here, exception should have been thrown.");
+    } catch (Exception e) {
+      assertTrue(e.getMessage().contains("Container checksum error"));
+    }
+  }
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/helpers/TestContainerUtils.java:144]: We already have this test case:

https://github.com/apache/ozone/blob/422df39669f424d62e98de061be6b90190907e91/hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestContainerDataYaml.java#L289-L301

New functionality is in `ContainerImporter#importContainer`, should be tested in `TestContainerImporter`.

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/helpers/TestContainerUtils.java [hunk 2]
```diff
@@ -122,6 +127,23 @@ public void testDatanodeIDPersistent(@TempDir File tempDir) throws Exception {
...
+    when(containerData.getContainerType()).thenReturn(
+        ContainerProtos.ContainerType.KeyValueContainer);
+    // mock to return different checksum
+    when(containerData.getChecksum()).thenReturn("checksum1", "checksum2");
+    conf.setBoolean(HddsConfigKeys.HDDS_CONTAINER_CHECKSUM_VERIFICATION_ENABLED, true);
+    try {
+      ContainerUtils.verifyChecksum(containerData, conf);
+      fail("code should not reach here, exception should have been thrown.");
+    } catch (Exception e) {
+      assertTrue(e.getMessage().contains("Container checksum error"));
```
Comment (jojochuang) [inline hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/helpers/TestContainerUtils.java:142]: use assertThrows to verify exception: https://www.digitalocean.com/community/tutorials/junit-assert-exception-expected

**Review by devmadhuu — COMMENTED** [review]
Thanks @DaveTeng0 for the patch. Just a minor comment.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ContainerImporter.java [hunk 3]
```diff
@@ -121,6 +125,7 @@ public void importContainer(long containerID, Path tarFilePath,
             .readContainer(containerDescriptorYaml);
       }
       containerData.setVolume(targetVolume);
+      ContainerUtils.verifyChecksum(containerData, conf);
```
[the remark this answers is not in the document]
Comment (DaveTeng0) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ContainerImporter.java:127]: yeah agree!

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/helpers/TestContainerUtils.java (hunk 2 above)
Comment (DaveTeng0) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/helpers/TestContainerUtils.java:142]: yeah agree!

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/helpers/TestContainerUtils.java (hunk 1 above)
Comment (DaveTeng0) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/helpers/TestContainerUtils.java:144]: yeah agree! updated in TestContainerImporter.

**Review by jojochuang — APPROVED** [review]

**Comment by jojochuang** [comment]
Thanks @DaveTeng0 @adoroszlai

**Pull request merged**